### PR TITLE
CIRC-1341: In-transit report performance: batch instances

### PR DESCRIPTION
**Pull request by yusuf-murodov-epam**
JIRA ticket [CIRC-1341](https://issues.folio.org/browse/CIRC-1341)

## Purpose
instances-in-transit report implementation. 

## Approach
Use the same fetcher as the original implementation to fetch instances in batches

### CODE REVIEW

File: src/main/java/org/folio/circulation/services/ItemsInTransitReportService.java [hunk 1]
```diff
@@ -79,9 +84,15 @@
...
   private CompletableFuture<Result<ItemsInTransitReportContext>> fetchInstances(
-    Result<ItemsInTransitReportContext> context) {
+    ItemsInTransitReportContext context) {
 
-    return completedFuture(context);
+    var mapper = new InstanceMapper();
+    FindWithMultipleCqlIndexValues<Instance> fetcher
+      = findWithMultipleCqlIndexValues(instancesStorageClient, "instances", mapper::toDomain);
+
+    return fetcher.findByIds(context.getItems().keySet())
```
Comment (viktor-draban) [inline src/main/java/org/folio/circulation/services/ItemsInTransitReportService.java:93]: So, you are looking for instances by their ids if I'm not mistaken of course but sent to this method itemIds (context.getItems().keySet()), right ?

Comment (yusuf-murodov-epam) [reply src/main/java/org/folio/circulation/services/ItemsInTransitReportService.java]: right, fixed. Thanks

File: src/main/java/org/folio/circulation/services/ItemsInTransitReportService.java [hunk 2]
```diff
@@ -79,9 +86,22 @@
...
+    var mapper = new InstanceMapper();
+    FindWithMultipleCqlIndexValues<Instance> fetcher
+      = findWithMultipleCqlIndexValues(instancesStorageClient, "instances", mapper::toDomain);
+
+    List<String> instanceIds = context.getItems().values()
+      .stream()
+      .map(Item::getInstanceId)
+      .filter(Objects::nonNull)
+      .distinct()
+      .collect(Collectors.toList());
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulation/services/ItemsInTransitReportService.java:100]: We'll need to do this quite often in this method - get some map from context, get some ids, remove nulls and collect distinct values. Can this be extracted to a method?

Comment (yusuf-murodov-epam) [reply src/main/java/org/folio/circulation/services/ItemsInTransitReportService.java]: I updated the code

File: src/main/java/org/folio/circulation/services/ItemsInTransitReportService.java [hunk 3]
```diff
@@ -79,9 +82,19 @@
   }
 
   private CompletableFuture<Result<ItemsInTransitReportContext>> fetchInstances(
-    Result<ItemsInTransitReportContext> context) {
+    ItemsInTransitReportContext context) {
 
-    return completedFuture(context);
+    return findInstanceIdsToFetch(context)
+      .after(instanceIds -> itemRepository.findInstancesByIds(instanceIds))
```
[the remark this answers is not in the document]
Comment (yusuf-murodov-epam) [reply src/main/java/org/folio/circulation/services/ItemsInTransitReportService.java]: fixed

File: src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java [hunk 4]
```diff
@@ -246,19 +254,24 @@ private ItemRepository(Clients clients, LocationRepository locationRepository,
...
           .map(item -> item.withInstance(mapper.toDomain(
               findById(item.getInstanceId(), instances.getRecords()).orElse(null))))
           .collect(Collectors.toList())));
     });
   }
 
+  private CompletableFuture<Result<MultipleRecords<JsonObject>>> fetchInstancesByIds(Collection<String> ids) {
+
+    final FindWithMultipleCqlIndexValues<JsonObject> fetcher
+      = findWithMultipleCqlIndexValues(instancesClient, "instances", identity());
```
[the remark this answers is not in the document]
Comment (yusuf-murodov-epam) [reply src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java]: fixed

[the remark this answers is not in the document]
Comment (yusuf-murodov-epam) [reply src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java]: fixed

File: src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java [hunk 5]
```diff
@@ -246,19 +254,24 @@ private ItemRepository(Clients clients, LocationRepository locationRepository,
...
+      return fetchInstancesByIds(instanceIds)
         .thenApply(r -> r.map(instances -> items.stream()
           .map(item -> item.withInstance(mapper.toDomain(
               findById(item.getInstanceId(), instances.getRecords()).orElse(null))))
           .collect(Collectors.toList())));
     });
   }
 
+  private CompletableFuture<Result<MultipleRecords<JsonObject>>> fetchInstancesByIds(Collection<String> ids) {
+
```
[the remark this answers is not in the document]
Comment (yusuf-murodov-epam) [reply src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java:269]: yes, it is more than 100 lines

File: src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java [hunk 6]
```diff
@@ -236,6 +237,13 @@ private ItemRepository(Clients clients, LocationRepository locationRepository,
         .collect(Collectors.toList()));
   }
 
+  public CompletableFuture<Result<MultipleRecords<Instance>>> findInstancesByIds(
+    Collection<String> ids) {
+
+    final var mapper = new InstanceMapper();
+    return fetchInstancesByIds(ids).thenApply(r -> r.map(multipleRecords -> multipleRecords.mapRecords(mapper::toDomain)));
```
[the remark this answers is not in the document]
Comment (yusuf-murodov-epam) [reply src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java]: fixed

File: src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java [hunk 7]
```diff
@@ -236,6 +237,13 @@ private ItemRepository(Clients clients, LocationRepository locationRepository,
         .collect(Collectors.toList()));
   }
 
+  public CompletableFuture<Result<MultipleRecords<Instance>>> findInstancesByIds(
+    Collection<String> ids) {
+
+    final var mapper = new InstanceMapper();
```
[the remark this answers is not in the document]
Comment (yusuf-murodov-epam) [reply src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java]: fixed

File: src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (roman-barannyk) [reply src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java:269]:
```suggestion
private CompletableFuture<Result<MultipleRecords<JsonObject>>> fetchInstancesByIds(
  Collection<String> ids) {
```

[the remark this answers is not in the document]
Comment (yusuf-murodov-epam) [reply src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java:270]: updated

File: src/main/java/org/folio/circulation/services/ItemsInTransitReportService.java [hunk 8]
```diff
@@ -115,9 +128,17 @@
...
-    Result<ItemsInTransitReportContext> context) {
+    ItemsInTransitReportContext context) {
 
-    return completedFuture(context);
+    final Result<CqlQuery> cqlQuery = exactMatchAny("status", openStates());
+    FindWithMultipleCqlIndexValues<Request> fetcher
+      = findWithMultipleCqlIndexValues(requestsStorageClient, "requests",
+      Request::from);
+
+    return fetcher.findByIdIndexAndQuery(context.getItems().keySet(), "itemId", cqlQuery)
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulation/services/ItemsInTransitReportService.java:138]: Why are we fetching requests again? They should be already fetched and put in the context by this point (fetchRequests(Result<ItemsInTransitReportContext> context))

**Review by alexanderkurash — APPROVED** [review]

File: src/main/java/org/folio/circulation/infrastructure/storage/users/UserRepository.java [hunk 9]
```diff
@@ -12,6 +12,7 @@
 import java.lang.invoke.MethodHandles;
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
```
Comment (sherzodn) [inline src/main/java/org/folio/circulation/infrastructure/storage/users/UserRepository.java:15]: Remove this unused import

Comment (yusuf-murodov-epam) [reply src/main/java/org/folio/circulation/infrastructure/storage/users/UserRepository.java]: removed

File: src/main/java/org/folio/circulation/services/ItemsInTransitReportService.java (hunk 8 above)
Comment (yusuf-murodov-epam) [reply src/main/java/org/folio/circulation/services/ItemsInTransitReportService.java]: yeah, it was redundant, fixed

File: src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java [hunk 10]
```diff
@@ -246,19 +257,22 @@ private ItemRepository(Clients clients, LocationRepository locationRepository,
...
           .map(item -> item.withInstance(mapper.toDomain(
               findById(item.getInstanceId(), instances.getRecords()).orElse(null))))
           .collect(Collectors.toList())));
     });
   }
 
+  private CompletableFuture<Result<MultipleRecords<JsonObject>>> fetchInstancesByIds(
+    Collection<String> ids) {
+
+    return findWithMultipleCqlIndexValues(instancesClient, INSTANCES_RECORDS_PROPERTY_NAME, identity()).findByIds(ids);
```
Comment (roman-barannyk) [inline src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java:273]: Seems like it's more than 100 chars again

File: src/test/java/org/folio/circulation/infrastructure/storage/inventory/UserRepositoryTest.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import io.vertx.core.json.JsonObject;
+
+/**
+ * Temporary tests to pass Sonar checks
+ * TODO: delete
```
Comment (sherzodn) [inline src/test/java/org/folio/circulation/infrastructure/storage/inventory/UserRepositoryTest.java:32]: why do you want delete this test?

Comment (yusuf-murodov-epam) [reply src/test/java/org/folio/circulation/infrastructure/storage/inventory/UserRepositoryTest.java:32]: because it is just to pass Sonar coverage test

Comment (sherzodn) [reply src/test/java/org/folio/circulation/infrastructure/storage/inventory/UserRepositoryTest.java:32]: I think you should not delete it, just keep it

File: src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java [hunk 12]
```diff
@@ -70,6 +72,8 @@
   private final boolean fetchLoanType;
 
   private static final String ITEMS_COLLECTION_PROPERTY_NAME = "items";
+  private static final String HOLDINGS_RECORDS_COLLECTION_PROPERTY_NAME = "holdingsRecords";
+  private static final String INSTANCES_RECORDS_PROPERTY_NAME = "instances";
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java:76]: INSTANCES_RECORDS_COLLECTION_PROPERTY_NAME?
Please check 100 char limit after renaming

File: src/test/java/org/folio/circulation/infrastructure/storage/inventory/UserRepositoryTest.java (hunk 11 above)
Comment (yusuf-murodov-epam) [reply src/test/java/org/folio/circulation/infrastructure/storage/inventory/UserRepositoryTest.java]: fixed

**Review by sherzodn — APPROVED** [review]

File: src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java (hunk 12 above)
Comment (yusuf-murodov-epam) [reply src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java:76]: just wanted to replace records to collection, fixed

File: src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java [hunk 13]
```diff
@@ -246,19 +258,23 @@ private ItemRepository(Clients clients, LocationRepository locationRepository,
...
               findById(item.getInstanceId(), instances.getRecords()).orElse(null))))
           .collect(Collectors.toList())));
     });
   }
 
+  private CompletableFuture<Result<MultipleRecords<JsonObject>>> fetchInstancesByIds(
+    Collection<String> ids) {
+
+    return findWithMultipleCqlIndexValues(instancesClient, INSTANCES_RECORDS_PROPERTY_NAME,
+      identity()).findByIds(ids);
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java:275]: Given that `INSTANCES_RECORDS_PROPERTY_NAME` is used only once in this private method, could this be a constant in the method (or even inline) in order to keep it within the same scope as it's usage (and make refactoring easier)?

Comment (yusuf-murodov-epam) [reply src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java]: fixed, I just updated method by merging and forgot to make it inline, thanks.

Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java]: Thank you. 

Can a similar refactoring be done for the `HOLDINGS_RECORDS_COLLECTION_PROPERTY_NAME`?

Comment (yusuf-murodov-epam) [reply src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java]: It is used more than once actually and seems not good idea to make inline on multiple places

Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java]: Agreed. 

My interpretation of other changes in this pull request was that the `fetchInstancesByID` method was extracted in order to reduce similar duplication for the instance fetcher. 

I wondered if the same could be done for the holdings fetcher.

If not, please resolve this conversation.

Comment (yusuf-murodov-epam) [reply src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java]: ok, thanks

Comment (yusuf-murodov-epam) [reply src/main/java/org/folio/circulation/infrastructure/storage/inventory/ItemRepository.java]: ok, extracted duplication part, and also made `HOLDINGS_RECORDS_PROPERTY_NAME` inline

**Pull request merged**